consensus/propagation/have_wants.go

```diff
@@ -63,58 +64,130 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
-		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
-			}
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
 	}
 
-	if hc.IsEmpty() {
-		return
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var want *proptypes.WantParts
+	batchRequestCount := int64(0)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case req, ok := <-ps.requestChan:
+			if !ok {
+				return
+			}
+
+			// concurrent per-peer per-block request limit reached
+			if ps.requestCount.Load()+batchRequestCount >= perPeerPerBlockConcurrentRequestLimit() {
+				// we reached the limit of requests, we request what we can
+				if want != nil && !want.Parts.IsEmpty() {
+					blockProp.sendWant(ps, want)
+					want = nil
+					batchRequestCount = 0
+				}
+				// wait for a part to be received before continuing
+				select {
+				case <-blockProp.ctx.Done():
+					return
+				case _, ok = <-ps.receivedPart:
+					if !ok {
+						return
+					}
+				}
+			}
+
+			// if we're at a new height, we can drop the previous requests
+			if want != nil &&
+				(want.Height != req.height || (want.Height == req.height && want.Round != req.round)) {
+				want = nil
+				batchRequestCount = 0
+			}
```

Comment:
why not just use the existing state for the relevant height and round? what advantage does having a laggy copy that's updated by what the peer sends us do better? just avoid a mutex call? I still feel like this is uneccesarily bug prone and we should change it if possible.

Reply:
how would we know that a bit array corresponds to which height? as described here: https://github.com/celestiaorg/celestia-core/pull/1773#discussion_r2061602078

Reply:
the purpose of the peer state is to keep track of state for each height and round that is relevant only for a specific peer, such as haves, wants, requests

we should use that